src/ServiceControl/Monitoring/HeartbeatMonitoring/HeartbeatTypesArrayToInstanceConverter.cs

```diff
@@ -0,0 +1,106 @@
+#nullable enable
+
+namespace ServiceControl.Monitoring.HeartbeatMonitoring
+{
+    using System;
+    using System.Text.Json;
+    using System.Text.Json.Serialization;
+    using Plugin.Heartbeat.Messages;
+    using ServiceControl.HeartbeatMonitoring.InternalMessages;
+
+    // ServiceControl.Plugin.Nsb5.Heartbeat used to send RegisterEndpointStartup and EndpointHeartbeat messages wrapped in an array.
+    // In order to stay backward compatible, we need to convert the array to an instance.
+    public class HeartbeatTypesArrayToInstanceConverter : JsonConverterFactory
+    {
+        public override bool CanConvert(Type typeToConvert) => typeToConvert == typeof(RegisterEndpointStartup) ||
+                                                               typeToConvert == typeof(EndpointHeartbeat) ||
+                                                               typeToConvert == typeof(RegisterPotentiallyMissingHeartbeats);
+
+        public override JsonConverter CreateConverter(Type typeToConvert, JsonSerializerOptions options) =>
+            typeToConvert switch
+            {
+                _ when typeToConvert == typeof(RegisterEndpointStartup) => new Converter<RegisterEndpointStartup>(options),
+                _ when typeToConvert == typeof(EndpointHeartbeat) => new Converter<EndpointHeartbeat>(options),
+                _ when typeToConvert == typeof(RegisterPotentiallyMissingHeartbeats) => new Converter<RegisterPotentiallyMissingHeartbeats>(options),
+                _ => throw new NotSupportedException()
+            };
+
+        sealed class Converter<TValue>(JsonSerializerOptions options) : JsonConverter<TValue>
+        {
+            // Currently we want to rely on deserializing the actual value by using the standard json serializer
+            // To make sure we are not recursively invoking the converter we need to remove it from the options
+            // Removing from the options is currently only possible when the converter is added to the options directly
+            // and not supported when the converter is added to the type
+            readonly JsonSerializerOptions optionsWithoutCustomConverter = options.FromWithout<HeartbeatTypesArrayToInstanceConverter>();
+
+            public override TValue? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            {
+                var readArray = false;
+                if (reader.TokenType == JsonTokenType.StartArray)
+                {
+                    readArray = true;
+                    reader.Read();
+
+                    if (reader.TokenType != JsonTokenType.StartObject)
+                    {
+                        throw new JsonException($"Unsupported structure for {typeToConvert}");
```

Comment:
See https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/converters-how-to?pivots=dotnet-8-0#jsonexception This change is unnecessary because the exception message doesn't provide more information than the built-in constructor unwinding

Reply:
That is nice! That actually is an amazing behavior I never really actively realized how that works. Will likely do that in personal parsing like projects in the future.

Will revert these based on that behavior.